src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,574 @@
+							  processing_config_size);
+		if (ret) {
+			comp_err(dev, "GoogleCtcAudioProcessingReconfigure failed: %d",
+				 ret);
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int ctc_prepare(struct processing_module *mod,
+		       struct sof_source **sources, int num_of_sources,
+		       struct sof_sink **sinks, int num_of_sinks)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct comp_dev *dev = mod->dev;
+	struct comp_buffer *source;
+	int ret;
+
+	comp_info(mod->dev, "ctc_prepare()");
+
+	source = list_first_item(&dev->bsource_list, struct comp_buffer, sink_list);
+	cd->ctc_func = ctc_find_func(audio_stream_get_frm_fmt(&source->stream));
+	if (!cd->ctc_func) {
+		comp_err(mod->dev, "ctc_prepare(): No suitable processing function found.");
+		ret = -EINVAL;
+		goto err;
+	}
+
+	return 0;
+err:
+	return ret;
+}
+
+static int ctc_reset(struct processing_module *mod)
+{
+	comp_dbg(mod->dev, "ctc_reset()");
+
+	return 0;
+}
+
+static int ctc_process(struct processing_module *mod,
+		       struct input_stream_buffer *input_buffers,
+		       int num_input_buffers,
+		       struct output_stream_buffer *output_buffers,
+		       int num_output_buffers)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	struct audio_stream *source = input_buffers[0].data;
+	struct audio_stream *sink = output_buffers[0].data;
+	uint32_t frames = input_buffers[0].size;
+
+	int ret;
+
+	comp_dbg(mod->dev, "ctc_process()");
+
+	if (cd->reconfigure) {
+		ret = google_ctc_audio_processing_reconfigure(mod);
+		if (ret)
+			return ret;
+	}
+
+	cd->ctc_func(cd, source, sink, &input_buffers[0], &output_buffers[0], frames);
+	return 0;
+}
+
+#if CONFIG_IPC_MAJOR_3
+static int ctc_cmd_set_data(struct processing_module *mod,
+			    struct sof_ipc_ctrl_data *cdata)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	int ret;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		ret = comp_data_blob_set_cmd(cd->tuning_handler, cdata);
+		if (ret)
+			return ret;
+		if (comp_is_new_data_blob_available(cd->tuning_handler)) {
+			comp_get_data_blob(cd->tuning_handler, NULL, NULL);
+			cd->reconfigure = true;
+		}
+		return 0;
+	default:
+		comp_err(mod->dev,
+			 "google_ctc_audio_processing_ctrl_set_data(): Only binary controls supported %d",
+			 cdata->cmd);
+		return -EINVAL;
+	}
+}
+
+static int ctc_cmd_get_data(struct processing_module *mod,
+			    struct sof_ipc_ctrl_data *cdata,
+			    size_t max_data_size)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	comp_info(mod->dev, "google_ctc_audio_processing_ctrl_get_data(): %u", cdata->cmd);
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		return comp_data_blob_get_cmd(cd->tuning_handler, cdata, max_data_size);
+	default:
+		comp_err(mod->dev,
+			 "google_ctc_audio_processing_ctrl_get_data(): Only binary controls supported %d",
+			 cdata->cmd);
```

Comment:
FWIW: that -EINVAL is propagated all the way out to the userspace process that sets the ALSA control, so it won't be missed.  Not clear to me that it's worth space in the firmware binary just to log "unsupported".  IMHO "ERROR" logs in firmware should be limited to genuine runtime errors, not routine checking of correct software usage.

Reply:
I agree. Thanks!